hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/MetricsBucket.java

```
@@ -0,0 +1,196 @@
+   * any remaining metrics.
+   * @param client the AbfsClient to deregister
+   * @return true if the client was deregistered, false otherwise
+   */
+  public boolean deregisterClient(AbfsClient client) {
+    if (client == null) {
+      return false;
+    }
+    ConcurrentLinkedQueue<String> batchToSend = null;
+    boolean isLastClient = false;
+
+    synchronized (this) {
+      if (!clients.contains(client)) {
+        return false;
+      }
+
+      if (clients.size() == 1) {
+        // This client is the last one — drain metrics now
+        batchToSend = metricsBuffer.getAndSet(new ConcurrentLinkedQueue<>());
+        isLastClient = true;
+      }
+
+      clients.remove(client);
+    }
+    if (isLastClient) {
+      sendMetrics(client, batchToSend);
+    }
+    return true;
+  }
+
+  /**
+   * Add a metric to the buffer.
+   * @param metric the metric to add
+   */
+  void addRequest(String metric) {
+    if (metric != null) {
+      metricsBuffer.get().add(metric);
+    }
+  }
+
+  /**
+   * Drain the metrics buffer and send if there are registered clients.
+   */
+  public void drainAndSendIfReady() {
+    AbfsClient client;
+
+    synchronized (this) {
+      if (clients.isEmpty()) {
+        return;
+      }
+      client = clients.iterator().next();
+    }
+
+    ConcurrentLinkedQueue<String> batch = metricsBuffer.getAndSet(
+        new ConcurrentLinkedQueue<>());
+    if (batch.isEmpty()) {
+      return;
+    }
+
+    sendMetrics(client, batch);
+  }
+
+  // Send metrics outside synchronized block
+  private void sendMetrics(AbfsClient client,
+      ConcurrentLinkedQueue<String> batchToSend) {
+    // Send outside synchronized block
+    if (client != null && batchToSend != null && !batchToSend.isEmpty()) {
+      for (String chunk : splitListBySize(batchToSend, MAX_HEADER_SIZE)) {
+        rateLimiter.acquire(); // Rate limiting
+        try {
+          client.getMetricCall(chunk);
+        } catch (IOException ignored) {}
```

Can add a debug line here instead of ignoring exception